File: sdk/ios/Sources/MessageBirdsAnalytics/EventQueue.swift
import Foundation

struct QueuedEvent: Codable {
    var envelope: EventEnvelope
    var attempts: Int
}

/// JSON-file-backed queue, mirroring `sdk/web`'s `EventQueue`
/// (`sdk/web/src/queue.ts`) design exactly: `POST /events` takes one
/// envelope, not a batch, so `flush` sends serially and stops at the
/// first failure rather than reordering or hammering a possibly-down
/// endpoint. An `actor` (not a class + locks) since queue mutation must be
/// serialized across concurrent `track`/`flush` calls.
actor EventQueue {
    private let fileURL: URL
    private let send: (EventEnvelope) async -> Bool
    private var nextRetryAt: Date = .distantPast
    private static let maxAttempts = 5

    init(fileURL: URL, send: @escaping (EventEnvelope) async -> Bool) {
        self.fileURL = fileURL
        self.send = send
    }

    func enqueue(_ envelope: EventEnvelope) {
        var items = load()
        items.append(QueuedEvent(envelope: envelope, attempts: 0))
        persist(items)
    }

    @discardableResult
    func flush(onDropped: ((EventEnvelope) -> Void)? = nil) async -> Bool {
        if Date() < nextRetryAt { return false }

        var items = load()
        var index = 0
        while index < items.count {
            let ok = await send(items[index].envelope)
            if ok {
                index += 1
                continue
            }

            items[index].attempts += 1
            if items[index].attempts >= Self.maxAttempts {
                onDropped?(items[index].envelope)
                index += 1
                continue
            }

            nextRetryAt = Date().addingTimeInterval(Self.backoffSeconds(attempts: items[index].attempts))
            persist(Array(items[index...]))
            return false
        }

        persist([])
        return true
    }

    private func load() -> [QueuedEvent] {
        guard let data = try? Data(contentsOf: fileURL) else { return [] }
        return (try? JSONDecoder().decode([QueuedEvent].self, from: data)) ?? []
    }

    private func persist(_ items: [QueuedEvent]) {
        guard let data = try? JSONEncoder().encode(items) else { return }
        try? data.write(to: fileURL, options: .atomic)
    }

    private static func backoffSeconds(attempts: Int) -> TimeInterval {
        min(30, pow(2, Double(attempts - 1)))
    }

    /// Test-only: bypasses the backoff delay so a retry test doesn't need
    /// to sleep for real. Not part of the SDK's real usage surface.
    func resetBackoffForTesting() {
        nextRetryAt = .distantPast
    }
}

File: sdk/android/src/main/kotlin/com/messagebirds/android/EventQueue.kt
package com.messagebirds.android

import java.io.File
import kotlin.math.min
import kotlin.math.pow
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import org.json.JSONArray
import org.json.JSONObject

private const val MAX_ATTEMPTS = 5

/**
 * JSON-file-backed queue, mirroring `sdk/web`'s `EventQueue`
 * (`sdk/web/src/queue.ts`) and `sdk/ios`'s `EventQueue`
 * (`sdk/ios/Sources/MessageBirdsAnalytics/EventQueue.swift`) exactly:
 * `POST /events` takes one envelope, not a batch, so `flush` sends
 * serially and stops at the first failure rather than reordering or
 * hammering a possibly-down endpoint. A `Mutex` serializes concurrent
 * `enqueue`/`flush` calls, the coroutine equivalent of Swift's `actor`.
 */
class EventQueue(
    private val file: File,
    private val send: suspend (EventEnvelope) -> Boolean,
) {
    private val mutex = Mutex()
    private var nextRetryAtMillis: Long = 0

    suspend fun enqueue(envelope: EventEnvelope) = mutex.withLock {
        val items = load().toMutableList()
        items.add(QueuedEvent(envelope, attempts = 0))
        persist(items)
    }

    /** @return true if the queue was fully drained. */
    suspend fun flush(onDropped: ((EventEnvelope) -> Unit)? = null): Boolean = mutex.withLock {
        if (System.currentTimeMillis() < nextRetryAtMillis) return@withLock false

        val items = load().toMutableList()
        var index = 0
        while (index < items.size) {
            val ok = send(items[index].envelope)
            if (ok) {
                index += 1
                continue
            }

            items[index] = items[index].copy(attempts = items[index].attempts + 1)
            if (items[index].attempts >= MAX_ATTEMPTS) {
                onDropped?.invoke(items[index].envelope)
                index += 1
                continue
            }

            nextRetryAtMillis = System.currentTimeMillis() + backoffMillis(items[index].attempts)
            persist(items.subList(index, items.size))
            return@withLock false
        }

        persist(emptyList())
        true
    }

    /** Test-only: bypasses the backoff delay so a retry test doesn't sleep for real. */
    internal suspend fun resetBackoffForTesting() = mutex.withLock {
        nextRetryAtMillis = 0
    }

    private fun load(): List<QueuedEvent> {
        if (!file.exists()) return emptyList()
        return try {
            val array = JSONArray(file.readText())
            (0 until array.length()).map { i ->
                val obj = array.getJSONObject(i)
                QueuedEvent(EventEnvelope.fromJson(obj.getJSONObject("envelope")), obj.getInt("attempts"))
            }
        } catch (e: Exception) {
            emptyList()
        }
    }

    private fun persist(items: List<QueuedEvent>) {
        val array = JSONArray()
        for (item in items) {
            array.put(
                JSONObject().apply {
                    put("envelope", item.envelope.toJson())
                    put("attempts", item.attempts)
                },
            )
        }
        file.writeText(array.toString())
    }

    private fun backoffMillis(attempts: Int): Long =
        min(30_000.0, 1_000.0 * 2.0.pow(attempts - 1)).toLong()
}

internal data class QueuedEvent(val envelope: EventEnvelope, val attempts: Int)
